File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        l, r = 0, len(height) - 1
        max_left, max_right = height[l], height[r]
        ans = 0
        while l < r:
            if height[l] <= height[r]:
                have = max_left - height[l]
                if height[l] > max_left:
                    max_left = height[l]
                ans += have if have > 0 else 0
                l += 1
            else:
                have = max_right - height[r]
                if height[r] > max_right:
                    max_right = height[r]
                ans += have if have > 0 else 0
                r -= 1
        return ans

        

        

File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int[] maxLeftHeight = new int[height.length];
        int[] maxRightHeight = new int[height.length];

        maxLeftHeight[0] = 0;
        maxRightHeight[height.length - 1] = 0;

        for(int i = 1; i < height.length; i++){
            maxLeftHeight[i] = Math.max(maxLeftHeight[i - 1], height[i - 1]);
        }

        for(int i = height.length - 2; i >= 0; i--){
            maxRightHeight[i] = Math.max(maxRightHeight[i + 1], height[i + 1]);
        }

        int ans = 0;
        for(int i = 0; i < height.length; i++){
            int trapped = Math.min(maxRightHeight[i], maxLeftHeight[i]) - height[i];
            if(trapped > 0){
                ans += trapped;
            }
        }
        return ans;
    }
}